### ISSUE TITLE
if dispatcher is run with token-enforcing regime, all public requests are forbidden. I guess we do not want that. we want to have both public (token-less) requests and authentified ones

### ISSUE THREAD

**Issue by volodymyrss** [oda-hub/dispatcher-app#12]

**Comment by volodymyrss** [comment]
create a configuration option whether token-less operation is allowed.
right now this is controlled with env variable

**Comment by burnout87** [comment]
Questions:

- Is the token part of the request received by the dispatcher?
- is it set on the frontend side?


Parametrized approach for all the requests:

- specify whether or not the request is public (default _yes_), this enables the token-less operation
- if otherwise, the request is not public, the token-check is enforced, and its presence and validity verified
- in this way the env variable is no longer necessary since this parameter will prevent the check of the token from happening when the request is public

**Comment by volodymyrss** [comment]
> Questions:
> 
>     * Is the token part of the request received by the dispatcher?
>     * is it set on the frontend side?

yes for both.

Note some description started [here](https://github.com/oda-hub/dispatcher-app/blob/master/interfaces.md) and is under review by [MM](https://github.com/orgs/oda-hub/people/motame)

> 
> 
> Parametrized approach for all the requests:
> 
>     * specify whether or not the request is public (default _yes_), this enables the token-less operation
> 
>     * if otherwise, the request is not public, the token-check is enforced, and its presence and validity verified
> 
>     * in this way the env variable is no longer necessary since this parameter will prevent the check of the token from happening when the request is public

sounds good, if I understood correctly. 



But indeed, the variable is just for the case when tokens are always needed. Note that in any case, right now, the  [function that validates token](https://github.com/oda-hub/dispatcher-app/blob/master/cdci_data_analysis/flask_app/dispatcher_query.py#L725) always returns true, hence it does not even matter how token is parsed, as long as it is possible to process request to token.

basically this function needs to convert token to list of roles, and if there is no token -  deduce empty list of roles.
then, it would need to check if roles are needed for some requests, as outlined in the [aforementioned doc](https://github.com/oda-hub/dispatcher-app/blob/master/interfaces.md).

**Comment by burnout87** [comment]
As described, a new parameter can be specified within the request. This is called _public_, and by default it is true. So when a token is sent this should be also set to False (or we can just decide to consider a request as non-public when the token is specified).

The next step will be to work on the validation of the token...or is there already something decided to be done on that side ?

Changes can be visible within the [https://github.com/oda-hub/dispatcher-app/tree/serverConfig-handling](url) branch, where a specific test is now available...but the validation of the token complete, more tests will of course be needed.

**Comment by volodymyrss** [comment]
> As described, a new parameter can be specified within the request. This is called _public_, and by default it is true. So when a token is sent this should be also set to False

> (or we can just decide to consider a request as non-public when the token is specified).

Maybe yes, else "public" seems a bit redundant, no?
presence/absence of of "token" carries the same information.
 
> The next steps will be:
> 
>     * work on the validation of the token...or is there already something decided to be done on that side ?

[there is a bit](https://github.com/oda-hub/dispatcher-app/blob/staging-1-3/interfaces.md#usage)

Please have a look if it makes sense.

>     * adapt the frontend so that the calls will now include this parameter

seems like Mohamed did it already (as [discussed](https://github.com/oda-hub/meetings/blob/main/2021-03-01/minutes.md#authentication) but I am not sure where the change is!

we should [discuss it next Monday](https://github.com/oda-hub/meetings/blob/main/2021-03-15/agenda.md)

And also [we need it in API](https://github.com/oda-hub/oda_api/issues/2), where it currently only passes it as a cookie - cookie is needed for OAuth2 proxy, and Bearer would be handled by the dispatcher. 
Maybe we can have an option how to provide token.

> 
> 
> Changes can be visible within the [https://github.com/oda-hub/dispatcher-app/tree/serverConfig-handling](url) branch, where a specific test is now available...but the validation of the token complete, more tests will of course be needed.

alright! By the way, you can already make a PR, keep it as Draft, or WIP.

**Comment by burnout87** [comment]
> > As described, a new parameter can be specified within the request. This is called _public_, and by default it is true. So when a token is sent this should be also set to False
> 
> > (or we can just decide to consider a request as non-public when the token is specified).
> 
> Maybe yes, else "public" seems a bit redundant, no?
> presence/absence of of "token" carries the same information.

Most likely I will go this way, it makes sense. A situation where a request where the token is not specified will be automatically considered as public, and therefore having the related privacy policies applied. 

Is a token anyway always sent from the frontend when the request is not public right?

> 
> > The next steps will be:
> > ```
> > * work on the validation of the token...or is there already something decided to be done on that side ?
> > ```
> 
> [there is a bit](https://github.com/oda-hub/dispatcher-app/blob/staging-1-3/interfaces.md#usage)
> 
> Please have a look if it makes sense.

Yes, I will look into it.

**Comment by volodymyrss** [comment]
> > Maybe yes, else "public" seems a bit redundant, no?
> > presence/absence of of "token" carries the same information.
> 
> Most likely I will go this way, it makes sense. A situation where a request where the token is not specified will be automatically considered as public, and therefore having the related privacy policies applied.

Ok!

It is also good that this approach is naturally compatible with current API version, so we do not need to increment it before it is really necessary, and all previous client versions will work.
 
> > [there is a bit](https://github.com/oda-hub/dispatcher-app/blob/staging-1-3/interfaces.md#usage)
> > Please have a look if it makes sense.
> 
> Yes, I will look into it.

probably, we could start to have a prototype following the mechanism described/agreed: with given responsibility on backends/plugins/frontend.

**Comment by burnout87** [comment]
Ok perfect.

In the meantime I asked @motame some token example so that I can start work on the token decoding and validation functionalities.

**Comment by volodymyrss** [comment]
ok. 
Part of the information is in [there](https://redmine.astro.unige.ch/issues/22906). 
But you need secret to validate.
Note that you can decode it without validating - it's not encrypted.

In python, you can use pyjwt to decode and/or validate.

**Comment by burnout87** [comment]
Exacly, now it is to be agreed on how the secret will be shared with the dispatcher.
